Tests/GRDBTests/FoundationDateTests.swift

```diff
@@ -179,7 +179,7 @@ class FoundationDateTests : GRDBTestCase {
             XCTAssertEqual(calendar.component(.hour, from: date), 1)
             XCTAssertEqual(calendar.component(.minute, from: date), 2)
             XCTAssertEqual(calendar.component(.second, from: date), 3)
-            XCTAssertTrue(abs(calendar.component(.nanosecond, from: date) - 4_000_000) < 10)  // We actually get 4_000_008. Some precision is lost during the DateComponents -> Date conversion. Not a big deal.
+            XCTAssertTrue(abs(calendar.component(.nanosecond, from: date) - 4_560_000) < 10)
```

Comment:
This change needs to be reverted.

The test was making sure that GRDB handles fractional seconds is a compatible way with SQLite. Extra precision has to be [ignored](https://github.com/groue/GRDB.swift/blob/v2.10.0/GRDB/Core/Support/Foundation/DatabaseDateComponents.swift#L268-L270):

```swift
let seconds = try Double.fetchOne(db, "SELECT strftime('%f','2015-07-22 01:02:03.00456')")!
print(seconds) // 3.005, not 3.00456
```

Reply:
@groue I'm already on it 😉 Another commit coming in a sec. 

Although SQLite *is* maintaining the precision for me locally. But we can't depend on that.

Reply:
> Although SQLite is maintaining the precision for me locally. But we can't depend on that.

Interesting. What do you mean?

Reply:
SQLiteDateParser is preserving precision to [9 sig figs](https://github.com/sobri909/GRDB.swift/blob/dad163771ff18368e54acd9c1248caab8ee94e66/GRDB/Core/Support/Foundation/SQLiteDateParser.swift#L101-L105). So that test is passing for me locally. (And on travis too, so far. Yay!)

Although sim tests != real device. And having the tests demand that level of precision would be brittle for any changes in SQLite engine. SQLite I don't think is making any promises on how much precision it maintains.

So I'm about to commit a change to the tests that permits precision loss down to 3 sig figs. (The old tests were *demanding* that precision loss, instead of just permitting it).

Reply:
@groue The tests were failing when precision was maintained, instead of merely permitting precision loss. So the tests were wrong 😉

The latest commit allows precision loss to the same degree as the Scanner based parser, and also passes when precision is maintained. 

SQLiteDateParser will maintain all the precision it's given, instead of discarding it.

Reply:
From the [SQLite docs](https://www.sqlite.org/lang_datefunc.html): 

> Exactly three digits are shown in the examples because only the first three digits are significant to the result, but the input string can have fewer or more than three digits and the date/time functions will still operate correctly. Similarly, format 12 is shown with 10 significant digits, but the date/time functions will really accept as many or as few digits as are necessary to represent the Julian day number.

Reply:
> @groue The tests were failing when precision was maintained, instead of merely permitting precision loss. So the tests were wrong 😉

You're right: "00456" was stripped to "004", when SQLite reads 0.005 fractional seconds, as shown above.

> The latest commit allows precision loss to the same degree as the Scanner based parser, and also passes when precision is maintained.

I understand you better know. But I don't agree with your "permit precision loss down to 3 sig figs" comment in https://github.com/groue/GRDB.swift/pull/334/commits/3bcc5b806425b3f94bc927e3360cdc566481c425. I want actual precision loss, not maximum precision loss.

Reply:
The previous tests would fail when precision was maintained (because `abs(4_560_000 - 4_000_000) > 10`). So the only way to pass the test would've been to discard the precision?

Reply:
I'm sorry for this nitpicking, especially since this PR changes code that had a bug, as you have well shown - I'm not here to give lessons. But I value above all *compatibility with SQLite* whenever we have to reproduce in Swift some inner SQLite algorithm.

Reply:
No worries 😄

I don't *think* there's any SQLite compatibility issues (unless I'm missing something - I'm not intimately familiar with SQLite yet). Their docs say they accept as many sig figs as given, and just don't promise to make use of any beyond three, when using `strftime()` etc.

So maintaining whatever precision we're given (and receive back) shouldn't create compatibility issues. I think?

Reply:
> The previous tests would fail when precision was maintained (because abs(4_560_000 - 4_000_000) > 10). So the only way to pass the test would've been to discard the precision?

The original test was:

```swift
// (parsing "2015-07-22 01:02:03.00456")
// We actually get 4_000_008. Some precision is lost during the DateComponents -> Date conversion. Not a big deal.
XCTAssertTrue(abs(calendar.component(.nanosecond, from: date) - 4_000_000) < 10)
```

So it was testing that GRDB 2.10 parses `.00456` into .004 (and strips the `56` digits)

This test was not correct, since SQLite parses `.00456` into .005 (as shown by `strftime('%f','2015-07-22 01:02:03.00456')`). There is some rounding here, not performed by GRDB 2.10.

**An ideal test would assert identical parsing between GRDB and SQLite.** In all rigor, I'm not yet sure of the correct way to reach this ideal 😅

Reply:
I think a serious issue is potential overflow [here](https://github.com/sobri909/GRDB.swift/blob/3bcc5b806425b3f94bc927e3360cdc566481c425/GRDB/Core/Support/Foundation/SQLiteDateParser.swift#L19). Because it's quite possible that someone could create date strings with more than 9 points of precision.

I'm going to fix that now.

Reply:
> since SQLite parses .00456 into .005

Only if passed through `strftime()`. The tests I changed are only testing storage and retrieval of `Date` instances. And SQLite preserves full precision in that case, because it's simply storing and returning strings.

Reply:
> So the only way to pass the test would've been to discard the precision?

Yes, yes: we'll have to discard precision as long as SQLite does.

Reply:
Except that SQLite is only discarding precision when manipulating dates, not when storing them. (Or at least, it's not discarding the precision when they're stored as strings). 

So there's no reason to discard the precision when doing storage and retrieval? There's only a need to be aware of potential precision loss, when manipulating the dates SQL-side.

Reply:
(our messages have overlapped, and it may look like I dismiss your answers - I don't)

> Their docs say they accept as many sig figs as given, and just don't promise to make use of any beyond three, when using strftime() etc.

OK, I'll look at that.

Reply:
Hah. No worries 😄

> this does not mean that I'll let GRDB drift from actual, observed SQLite behavior

Likewise! The docs do say that it's fine to store more precision, and that the date/time functions will simply ignore it. But it will be worth testing to make sure that's definitely true.